python/plugins/processing/gui/NumberInputPanel.py

```diff
@@ -74,15 +80,64 @@ def __init__(self, number, minimum, maximum, isInteger):
             self.spnValue.setValue(0)
             self.spnValue.setClearValue(0)
 
-        self.btnCalc.clicked.connect(self.showNumberInputDialog)
+        self.btnCalc.setFixedHeight(self.spnValue.height())
+
+        self.btnCalc.clicked.connect(self.showExpressionsBuilder)
 
         self.spnValue.valueChanged.connect(lambda: self.hasChanged.emit())
 
-    def showNumberInputDialog(self):
-        dlg = NumberInputDialog(self.isInteger)
-        dlg.exec_()
-        if dlg.value is not None:
-            self.spnValue.setValue(dlg.value)
+    def showExpressionsBuilder(self):
+        context = QgsExpressionContext()
+        context.appendScope(QgsExpressionContextUtils.globalScope())
+        scope = QgsExpressionContextUtils.projectScope()
+        self.addVariablesToScope(scope)
+        context.appendScope(scope)
+        dlg = QgsExpressionBuilderDialog(None, self.spnValue.text(), self, "generic", context)
+        dlg.setWindowTitle(self.tr("Expression based input"));
+        if dlg.exec_() == QDialog.Accepted:
+            exp = QgsExpression(dlg.expressionText())
+            if not exp.hasParserError():
+                result =  exp.evaluate(context)
+                if not exp.hasEvalError():
+                    try:
+                        self.spnValue.setValue(float(result))
+                    except:
+                        pass
+
+    def addVariablesToScope(self, scope):
+        layers = dataobjects.getAllLayers()
+        for layer in layers:
+            name = layer.name()
+            scope.setVariable("%s_minx" % name, layer.extent().xMinimum())
+            scope.setVariable("%s_miny" % name, layer.extent().yMinimum())
+            scope.setVariable("%s_maxx" % name, layer.extent().xMaximum())
+            scope.setVariable("%s_maxy" % name, layer.extent().yMaximum())
+            if isinstance(layer, QgsRasterLayer):
+                cellsize = (layer.extent().xMaximum()
+                            - layer.extent().xMinimum()) / layer.width()
+                scope.setVariable("%s_cellsize" % name, cellsize)
+
+        layers = dataobjects.getRasterLayers()
+        for layer in layers:
+            for i in range(layer.bandCount()):
+                stats = layer.dataProvider().bandStatistics(i + 1)
+                scope.setVariable("%s_band%i_avg" % (name, i + 1), stats.mean)
+                scope.setVariable("%s_band%i_stddev" % (name, i + 1), stats.stdDev)
+                scope.setVariable("%s_band%i_min" % (name, i + 1), stats.minimumValue)
+                scope.setVariable("%s_band%i_max" % (name, i + 1), stats.maximumValue)
+
+        extent = iface.mapCanvas().extent()
+        scope.setVariable("canvasextent_minx", extent.xMinimum())
```

Comment:
All these new variables should have a description added in QgsExpression::initVariableHelp()